src/saga_action_func.rs

```diff
@@ -135,125 +83,88 @@ where
      * interfaces of its own so there's generally no need to have the specific
      * type.)
      */
-    pub fn new_action(
+    pub fn new_action<UserType, ActionFuncOutput>(
         action_func: ActionFuncType,
         undo_func: UndoFuncType,
-    ) -> Arc<dyn Action<UserType>> {
-        Arc::new(ActionFunc { action_func, undo_func, phantom: PhantomData })
+    ) -> Arc<dyn Action<UserType>>
```

Comment:
This is interesting.  If I'm understanding this right, `ActionFn` is more general than it needs to be in that it has no constraint that the output impls `ActionData`.  Instead, this constraint is moved to this function.  A consequence is that there's no need for the PhantomData in `struct ActionFunc`.  Is that right?

I'm thinking about how I might prompt myself to consider this pattern in the future.  When I have some expectation about the trait bounds on a type parameter for a struct, I tend to write them down at least on the impl block.  I can see how that's not ideal in this case.  Is there a general lesson here?

Reply:
The `ActionData` constraint is exclusive to the "forward edge", right? It doesn't seem like the "backward edge" (undo) has any such constraint on the output.

I dunno if this is a general lesson, but it kinda seems like a matter of:
- Having a shared abstraction for edges (in `ActionFn`)
- Since the constraints are different for action/undo, they aren't in the shared abstraction
- Instead, those constraints are pushed out to where the action/undo distinction is made (which is here in `new_action`).

This constraint could *probably* exist within the `ActionFunc` struct, rather than the impl, but this would functionally be identical to having the constraint on the constructor (since there only is one constructor). Plus, without https://rust-lang.github.io/rfcs/2089-implied-bounds.html , it would need to be added to the constructor anyway.